Main.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Main where
+
+import           Control.Monad
+import           Control.Monad.ST
+import           Data.Char
+import           Data.Foldable
+import           Data.Functor
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as M
+import           Numeric
+import           System.Environment
+import           Text.Parsec.Prim
+import           Text.ParserCombinators.Parsec hiding (spaces, try)
+
+symbol :: Parser Char
+symbol = oneOf "!#$%&|*+-/:<=>?@^_~"
+
+readExpr :: String -> String
+readExpr input = case parse parseExpr "scheme" input of
+  Left err -> "No match: " ++ show err
+  Right val -> "Found value:\n" ++ show val
+
+spaces :: Parser ()
+spaces = skipMany1 space
+
+main :: IO ()
+main = do
+  (expr:_) <- getArgs
+  putStrLn (readExpr expr)
+
+data LispValue = Atom String
+               | List [LispValue]
+               | DottedList [LispValue] LispValue
+               | Vector (V.Vector LispValue)
+               | RealNumber LispReal
+               | RationalNumber LispRational
+               | ComplexNumber LispComplex
+               | Character String
+               | String String
+               | Bool Bool
+               deriving (Eq, Show)
+
+data LispComplex =
+  LispComplex LispReal Char LispReal
```

Comment:
It might be better to use `Bool` instead of `Char` or even create a special `data Sign = Positive | Negative`